esp-link/cgimega.c

```diff
@@ -0,0 +1,1384 @@
+    }
+  } else if (len < 0) {
+    return len;	// This is bad but already reported this, so just pass the info.
+  } else {
+    os_printf("Sync reply length invalid : %d should be 11\n", len);
+    return len;
+  }
+}
+
+static void ICACHE_FLASH_ATTR sendRebootMCUQuery() {
+  pbuf.ms1 = 0;
+  pbuf.ms2 = 3;
+  pbuf.body[0] = CMD_LEAVE_PROGMODE_ISP;
+  pbuf.body[1] = 1;
+  pbuf.body[2] = 1;
+
+  writePacket();
+}
+
+static void ICACHE_FLASH_ATTR readRebootMCUReply() {
+  int len = readPacket();
+  if (len == 2) {
+    if (pbuf.body[0] == CMD_LEAVE_PROGMODE_ISP && pbuf.body[1] == STATUS_CMD_OK) {
+      reply_ok = true;
+      if (debug()) os_printf("Rebooting MCU : ok\n");
+      return;
+    }
+  }
+}
+
+static void ICACHE_FLASH_ATTR sendReadFlashQuery() {
+  pbuf.ms1 = 0;
+  pbuf.ms2 = 4;
+  pbuf.body[0] = CMD_READ_FLASH_ISP;
+  pbuf.body[1] = 0;	// Number of bytes to read, MSB first
+  pbuf.body[2] = 16;
+  pbuf.body[3] = 0x20;	// ??
+
+  writePacket();
+}
+
+static void ICACHE_FLASH_ATTR readReadFlashReply(char *buf) {
+  int len = readPacket();
+  if (len == 19) {
+    if (pbuf.body[0] == CMD_READ_FLASH_ISP && pbuf.body[1] == STATUS_CMD_OK) {
+      int i;
+      for (i=0; i<len-3; i++)
+        buf[i] = pbuf.body[i+2];
+    }
+    reply_ok = true;
+  } else
+    reply_ok = false;
+}
+
+static void ICACHE_FLASH_ATTR sendLoadAddressQuery(uint32_t addr) {
+  pbuf.ms1 = 0;
+  pbuf.ms2 = 5;
+
+  // Divide address by 2, to cope with addressing words, not bytes
+  int a2 = addr >> 1;
+
+  if (pbuf.body == 0)
+    return;
+  pbuf.body[0] = CMD_LOAD_ADDRESS;
+  pbuf.body[1] = 0xFF & (a2 >> 24);
+  pbuf.body[2] = 0xFF & (a2 >> 16);
+  pbuf.body[3] = 0xFF & (a2 >> 8);
+  pbuf.body[4] = 0xFF & a2;
+
+  writePacket();
+}
+
+static void ICACHE_FLASH_ATTR readLoadAddressReply() {
+  reply_ok = false;
+
+  int len = readPacket();
+  if (len == 2) {
+    if (pbuf.body[0] == CMD_LOAD_ADDRESS && pbuf.body[1] == STATUS_CMD_OK) {
+      reply_ok = true;
+      // os_printf("LoadAddress Reply : ok\n");
+      return;
+    }
+  }
+
+  os_printf("LoadAddress : %02x %02x (expected %02x %02x), len %d exp 2\n",
+    pbuf.body[0], pbuf.body[1], CMD_LOAD_ADDRESS, STATUS_CMD_OK, len);
+}
+
+static void ICACHE_FLASH_ATTR sendProgramPageQuery(char *data, int pgmLen) {
+  int totalLen = pgmLen + 10;
+  int mode = 0xC1;		// already includes the 0x80 "write page" bit
+
+  pbuf.ms1 = totalLen >> 8;
+  pbuf.ms2 = totalLen & 0xFF;
+
+  pbuf.body[0] = CMD_PROGRAM_FLASH_ISP;
+  pbuf.body[1] = pgmLen >> 8;
+  pbuf.body[2] = pgmLen & 0xFF;
+  pbuf.body[3] = mode;		// mode byte
+  pbuf.body[4] = 0x0A;		// delay
+  pbuf.body[5] = 0x40;		// cmd1
+  pbuf.body[6] = 0x4C;		// cmd2
+  pbuf.body[7] = 0x20;		// cmd3
+  pbuf.body[8] = 0x00;		// poll1
+  pbuf.body[9] = 0x00;		// poll2
+  for (int i=0, j=10; i<pgmLen; i++, j++)
+    pbuf.body[j] = data[i];
+
+      os_delay_us(4500L);	// Flashing takes about 4.5ms
+  writePacket();
+      os_delay_us(4500L);	// Flashing takes about 4.5ms
```

Comment:
why are there two delays? and why delay as opposed to polling for an ack with a timeout?